pkg/audit/verifier.go

```diff
@@ -196,23 +197,65 @@ func calcPadded(size int64, blockSize int) int64 {
 }
 
 // verify downloads shares then verifies the data correctness at the given stripe
-func (verifier *Verifier) verify(ctx context.Context, stripeIndex int, pointer *pb.Pointer) (failedNodes []*pb.Node, err error) {
+func (verifier *Verifier) verify(ctx context.Context, stripeIndex int, pointer *pb.Pointer) (verifiedNodes []*proto.Node, err error) {
 	defer mon.Task()(&ctx)(&err)
 
 	shares, nodes, err := verifier.downloader.DownloadShares(ctx, pointer, stripeIndex)
 	if err != nil {
 		return nil, err
 	}
 
+	var offlineNodes []string
+	for i, share := range shares {
+		if shares[i].Error != nil {
+			offlineNodes = append(offlineNodes, nodes[share.PieceNumber].GetId())
+		}
+	}
+
 	required := int(pointer.Remote.Redundancy.GetMinReq())
 	total := int(pointer.Remote.Redundancy.GetTotal())
 	pieceNums, err := auditShares(ctx, required, total, shares)
 	if err != nil {
 		return nil, err
 	}
 
+	var failedNodes []string
 	for _, pieceNum := range pieceNums {
-		failedNodes = append(failedNodes, nodes[pieceNum])
+		failedNodes = append(failedNodes, nodes[pieceNum].GetId())
+	}
+
+	successNodes := setSuccessNodes(ctx, nodes, failedNodes, offlineNodes)
+	verifiedNodes = setVerifiedNodes(ctx, nodes, offlineNodes, failedNodes, successNodes)
+
+	return verifiedNodes, nil
+}
+
+func setSuccessNodes(ctx context.Context, nodes []*pb.Node, failedNodes, offlineNodes []string) (successStatusNodes []*proto.Node) {
+	fails := make(map[string]bool)
+	for _, fail := range failedNodes {
+		fails[fail] = true
 	}
-	return failedNodes, nil
+	for _, offline := range offlineNodes {
+		fails[offline] = true
+	}
+
+	var successNodes []string
+	for _, node := range nodes {
+		if !fails[node.GetId()] {
+			successNodes = append(successNodes, node.GetId())
+		}
+	}
+	successStatusNodes = setSuccessStatus(ctx, successNodes)
+	return successStatusNodes
+}
+
+func setVerifiedNodes(ctx context.Context, nodes []*pb.Node, offlineNodes, failedNodes []string, successNodes []*proto.Node) (verifiedNodes []*proto.Node) {
```

Comment:
Just for consistency I would argue that you do the `setSuccessStatus` step in this function, rather than in `setSuccessNodes`. I'd rename `setSuccessNodes` to `getSuccessNodes` and have it return a string slice of node IDs, just like `offlineNodes` and `failedNodes`. Then you take in three string arrays of Node IDs and all the other logic stays the same.

Reply:
that's fair, I forget my rationale for having it differently lol